src/components/transport_manager/src/tcp/tcp_client_listener.cc

```diff
@@ -180,8 +180,45 @@ void SetKeepaliveOptions(const int fd) {
 #endif  // __QNX__
 }
 
+void removeFinalizedConnections(const TransportAdapterController* controller,
+                                ConnectionHolder& connections) {
+  ConnectionHolder::iterator connection_iter = connections.begin();
+  while(connection_iter != connections.end()) {
+    ConnectionHolderIndex current_index = connection_iter->first;
+    ConnectionSPtr current_connection = connection_iter->second;
+    if (controller->IsConnectionFinished(current_index.first,
+                                         current_index.second)) {
+      current_connection->Disconnect();
+      connections.erase(connection_iter++);
+    } else {
+      ++connection_iter;
+    }
+  }
+}
+
+void addConnection(const int connection_fd,
+                   TransportAdapterController* controller,
+                   TcpDevice* tcp_device, ConnectionHolder& connections) {
+  const ApplicationHandle app_handle = tcp_device->AddIncomingApplication(
+      connection_fd);
+  const DeviceUID& device_id = tcp_device->unique_device_id();
+  const ConnectionHolderIndex connection_index(device_id, app_handle);
+
+  if (connections.end() == connections.find(connection_index)) {
+    TcpSocketConnection* connection(
+        new TcpSocketConnection(device_id, app_handle, controller));
+    ConnectionSPtr tcp_connection(connection);
```

Comment:
@vlantonov Why don't you create SharedPtr at once instead of creating raw pointer than wrapping it to SharedPtr? 
I mean something like this: ConnectionSPtr tcp_connection(MakeShared<TcpSocketConnection>(device_id, app_handle, controller))